src/NHibernate/Engine/ActionQueue.cs

```diff
@@ -125,34 +136,58 @@ private void ExecuteActions(IList list)
 			// It will then fail here due to list being modified. (Some previous code was dodging the
 			// trouble with a for loop which was not failing provided the list was not getting smaller.
 			// But then it was clearing it without having executed added actions (if any), ...)
+			
 			foreach (IExecutable executable in list)
-				Execute(executable);
-
+			{
+				InnerExecute(executable);
+			}
 			list.Clear();
 			session.Batcher.ExecuteBatch();
+			
+		}
+
+		private void AfterExecutions()
+		{
+			if (session.Factory.Settings.IsQueryCacheEnabled)
+			{
+				var spaces = executedSpaces.ToArray();
+				afterTransactionProcesses.AddSpacesToInvalidate(spaces);
```

Comment:
I would make the parameter of `AddSpacesToInvalidate` `IReadOnlyCollection<string>` to avoid calling `.ToArray()`. 

Or, actually, remove that method and use `executedSpaces` directly instead of `querySpacesToInvalidate` collection.